flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java

```diff
@@ -0,0 +1,214 @@
+            if (throwable instanceof SavepointCreationException) {
+                throw (SavepointCreationException) throwable;
+            } else {
+                throw new CompletionException(
+                        "Error while stopping job after creating savepoint", throwable);
+            }
+        }
+
+        if (statusAndPath.getStatus() == JobStatus.FINISHED) {
+            // ensure that all vertices from the execution graph are in state FINISHED. There might
+            // be cases where only some subgraphs of the job are finished.
+            if (!areAllVerticesFinished()) {
+                if (getLogger().isDebugEnabled()) {
+                    getLogger().debug("Printing vertex states:");
+                    for (Execution e : getExecutionGraph().getRegisteredExecutions().values()) {
+                        getLogger().debug("{} = {}", e.getVertex(), e.getState());
+                    }
+                }
+                final FlinkException exception =
+                        new FlinkException(
+                                "Not all vertices of job are finished. Stop with savepoint operation failed.");
+                context.runIfState(this, () -> handleAnyFailure(exception));
+                throw new CompletionException("Unable to stop with savepoint", exception);
+            }
+            context.runIfState(
+                    this,
+                    () ->
+                            context.goToFinished(
+                                    ArchivedExecutionGraph.createFrom(getExecutionGraph())));
+            return statusAndPath.getPath();
+        } else {
+            throw new CompletionException(
+                    new FlinkException(
+                            "Reached state "
+                                    + statusAndPath.getStatus()
+                                    + " instead of FINISHED "));
+        }
+    }
+
+    private boolean areAllVerticesFinished() {
+        return getExecutionGraph().getRegisteredExecutions().values().stream()
+                .noneMatch(e -> e.getState() != ExecutionState.FINISHED);
+    }
+
+    @Override
+    protected void handleAnyFailure(@Nonnull Throwable cause) {
+        Preconditions.checkState(context.isState(this), "Assuming StopWithSavepoint state");
+        // restart the checkpoint coordinator if stopWithSavepoint failed.
+        startCheckpointScheduler(checkpointCoordinator);
+
+        if (getLogger().isDebugEnabled()) {
+            getLogger()
+                    .debug(
+                            "Restarting or Failing job because of failure while stopping with savepoint.",
+                            cause);
+        } else {
+            getLogger()
+                    .info(
+                            "Restarting or Failing job because of failure while stopping with savepoint. Reason: {}",
+                            cause.getMessage());
+        }
+        super.handleAnyFailure(cause);
```

Comment:
This is a bit of personal taste but I find it hard to see what happens in case of a failure in the state `StopWithSavepoint`. If we didn't use the `ExecutingStateWithFailureHandler` and instead have an explicit `context.goToRestarting` then it would be a bit clearer. It currently works because all failures are handled as global failovers. But once this changes, this class will break because we want to fail with global failovers explicitly.